docs/development/pigs/pig-009.rst

```diff
@@ -0,0 +1,298 @@
+allows to sample from it with the `.sample()` method:
+
+.. code::
+
+    lightcurve = LightCurveTableModel() # or PhaseCurveTableModel()
+    src_sampler = MapEventSampler(npred_map, lightcurve, t_min=, t_max=)
+
+    # draw total number of events
+    n_events = src_sampler.npred_total()
+
+    # draw events as an astropy table
+    events_src = map_sampler.sample()
+
+If no lightcurve is provided a constant rate can be assumed e.g. for 
+background models:
+
+.. code::
+
+    bkg_map = BackgroundModel().map
+    bkg_sampler = MapEventSampler(bkg_map, t_min=, t_max=)
+    events_bkg = bkg_sample.sample()
+
+
+The `MapEventSampler` returns an `astropy.table.Table` object, with the following
+columns: `TIME`, `RA_TRUE`, `DEC_TRUE`, `ENERGY_TRUE`. Later additional info such
+as `DETX` and `DETY` could be added.
+
+
+IRFEventDistributor
+-------------------
+
+We also propose a class to handle correctly the spatially varying IRFs: the
+`IRFEventDistibutor` class. Based on the `PSFMap` and `EDispMap`
+as an input it interpolates the "correct" IRF at the position of a given event and
+applies it. In more detail the class will calculate the psf and the energy dispersion 
+for each event true position and true energy. The IRFs are assumed to be constant
+and not time-dependent. The time dependency will be handled higher level, by a loop
+over good time-intervals (GTIs). The class would be used as following:
+
+.. code::
+
+    events = vstack([events_src_1, events_src_2, ..., events_src_N])
+    event_distributor = IRFEventDistributor(psf_map, edisp_map)
+
+    events = irf_distributor.sample_psf(events)
+    events = irf_distributor.sample_edisp(events)
+
+Alternatively the IRFs could be applied per model component and be handled in 
+the `MapEventSampler`. However in this case the IRFs would not be applied 
+correctly for e.g. very extended sources, because the IRF changes over region 
+of the source. If we find the approach with the `IRFEventDistibutor` to be not 
+performant enough or problematic in memory use, it can be still used as a
+simple fallback solution.
+
+The `IRFEventDistributor` adds the following columns to the event table:
+`RA`, `DEC` and `ENERGY`.
+
+
+MapDatasetEventSampler
+----------------------
+
+To combine the `MapEventSampler` and `IRFEventDistibutor` objects we finally propose 
+to introduce a `MapDatasetEventSampler`, which executes the sampling from all 
+source components, applies the IRFs, samples from the background model and 
+returns a stacked, combined event list.
+
+This is the proposed high level API:
+
+.. code::
+    
+    dataset_sampler = MapDatasetEventSampler(dataset, pointing)
+    events = dataset_sampler.sample()
+
+This requires to add `t_min` and `t_max` information to `MapDataset` or maybe a 
+`MapDataset.gti` attribute, defining the GTI for the given dataset. In addition 
+it also requires to add a lightcurve model to the `SkyModel`. Alternatively the
```

Comment:
"Alternatively ..."

You can't have both. You have to decide what you take in `__init__`.

If useful, you can use `@classmethod`s to have multiiple constructors.

Please make a choice for the main proposal, and (only if useful) - describe alternatives in the alternatives section.